www/js/lib/filecache.js

```diff
@@ -0,0 +1,221 @@
+        if (this._first === null) {
+            this._first = this._last = entry;
+        } else {
+            this._first.prev = entry;
+            entry.next = this._first;
+            this._first = entry;
+        }
+    };
+
+    /**
+     * Move a cache entry to the top of the cache
+     * @param {CacheEntry} entry The entry to move 
+     */
+    LRUCache.prototype.moveToTop = function (entry) {
+        this.unlink(entry);
+        this.insertAtTop(entry);
+    };
+
+    /**
+     * A new Block Cache
+     * @type {BlockCache}
+     */
+    var cache = new LRUCache();
+
+    // Counters for reporting only
+    var hits = 0;
+    var misses = 0;
+
+    /**
+     * Initializes or resets the cache - this should be called whenever a new ZIM is loaded
+     */
+    var init = function () {
+        console.log('Initialize or reset FileCache');
+        cache._entries = new Map();
+        // Initialize linked list of entries
+        cache._first = null;
+        cache._last = null;
+    };
+
+    /**
+     * Read a certain byte range in the given file, breaking the range into chunks that go through the cache
+     * If a read of more than BLOCK_SIZE * 2 (bytes) is requested, do not use the cache
+     * @param {Object} file The requested ZIM archive to read from
+     * @param {Integer} begin The byte from which to start reading
+     * @param {Integer} end The byte at which to stop reading (end will not be read)
+     * @return {Promise<Uint8Array>} A Promise that resolves to the correctly concatenated data from the cache 
+     *     or from the ZIM archive
+     */
+    var read = function (file, begin, end) {
+        // Read large chunks bypassing the block cache because we would have to
+        // stitch together too many blocks and would clog the cache
+        if (end - begin > BLOCK_SIZE * 2) return file._readSplitSlice(begin, end);
+        var readRequests = [];
+        var blocks = {};
+        // Look for the requested data in the blocks: we may need to stitch together data from two or more blocks
+        for (var id = Math.floor(begin / BLOCK_SIZE) * BLOCK_SIZE; id < end; id += BLOCK_SIZE) {
+            var block = cache.get(id + ':' + file.id);
+            if (block === undefined) {
+                // Data not in cache, so read from archive
+                misses++;
+                readRequests.push(function (offset) {
+                    return file._readSplitSlice(offset, offset + BLOCK_SIZE).then(function (result) {
+                        cache.store(offset + ':' + file.id, result);
+                        blocks[offset] = result;
+                    });
+                }(id));
+            } else {
+                hits++;
+                blocks[id] = block;
+            }
+        }
+        if (misses + hits > 2000) {
+            console.log("** Block cache hit rate: " + Math.round(hits / (hits + misses) * 1000) / 10 + "% [ hits:" + hits + " / misses:" + misses + " ]");
```

Comment:
I agree

Reply:
@mossroy I shall remove all console logging from this file before merging. However, I think it is woth keeping this line commented out, and the associated counters (`hits` and `misses`, also commented out) because future DEVs may wish to tune the Cache according to changing software and hardware environments.